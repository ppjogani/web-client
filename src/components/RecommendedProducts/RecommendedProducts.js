import React, { useEffect } from 'react';
import { connect } from 'react-redux';
import { compose } from 'redux';
import classNames from 'classnames';

import { FormattedMessage, useIntl } from '../../util/reactIntl';
import { propTypes } from '../../util/types';
import { formatMoney } from '../../util/currency';
import { createSlug } from '../../util/urlHelpers';
import { getMarketplaceEntities } from '../../ducks/marketplaceData.duck';
import { H3, NamedLink, ResponsiveImage } from '../../components';

import { fetchRecommendedProducts } from '../../ducks/recommendedProducts.duck';
import css from './RecommendedProducts.module.css';

/**
 * Simple product card for recommended products
 */
const RecommendedProductCard = ({ product, intl }) => {
  const { title, price, images } = product.attributes;
  const { sku } = product.attributes.publicData || {};

  const primaryImage = images && images.length > 0 ? images[0] : null;
  const formattedPrice = price ? formatMoney(intl, price) : null;
  const slug = createSlug(title);

  const linkProps = {
    name: 'ListingPage',
    params: { id: product.id.uuid, slug },
  };

  return (
    <NamedLink className={css.productCard} {...linkProps}>
      <div className={css.imageContainer}>
        {primaryImage ? (
          <ResponsiveImage
            rootClassName={css.productImage}
            alt={title}
            image={primaryImage}
            variants={['landscape-crop', 'landscape-crop2x']}
            sizes="(max-width: 767px) 100vw, (max-width: 1023px) 50vw, 33vw"
          />
        ) : (
          <div className={css.noImage}>
            <FormattedMessage id="RecommendedProducts.noImage" />
          </div>
        )}
      </div>
      <div className={css.productInfo}>
        <h4 className={css.productTitle}>{title}</h4>
        {formattedPrice && <div className={css.productPrice}>{formattedPrice}</div>}
      </div>
    </NamedLink>
  );
};

/**
 * RecommendedProducts component displays a "You may also like" section
 * with products fetched based on SKUs from publicData.recommendedProducts
 */
const RecommendedProductsComponent = props => {
  const {
    rootClassName,
    className,
    recommendedProductSKUs,
    recommendedProducts,
    fetchRecommendedProductsInProgress,
    fetchRecommendedProductsError,
    onFetchRecommendedProducts,
  } = props;

  const intl = useIntl();

  useEffect(() => {
    if (recommendedProductSKUs && recommendedProductSKUs.length > 0) {
      onFetchRecommendedProducts(recommendedProductSKUs);
    }
  }, [recommendedProductSKUs, onFetchRecommendedProducts]);

  // Don't render if no recommended product SKUs
  if (!recommendedProductSKUs || recommendedProductSKUs.length === 0) {
    return null;
  }

  const classes = classNames(rootClassName || css.root, className);
  const hasError = !!fetchRecommendedProductsError;
  const isLoading = fetchRecommendedProductsInProgress;
  const hasProducts = recommendedProducts && recommendedProducts.length > 0;

  return (
    <div className={classes}>
      <H3 as="h2" className={css.title}>
        <FormattedMessage id="RecommendedProducts.title" />
      </H3>

      {hasError ? (
        <div className={css.error}>
          <FormattedMessage id="RecommendedProducts.loadError" />
        </div>
      ) : isLoading ? (
        <div className={css.loading}>
          <FormattedMessage id="RecommendedProducts.loading" />
        </div>
      ) : hasProducts ? (
        <div className={css.productsGrid}>
          {recommendedProducts.map(product => (
            <RecommendedProductCard
              key={product.id.uuid}
              product={product}
              intl={intl}
            />
          ))}
        </div>
      ) : (
        <div className={css.noProducts}>
          <FormattedMessage id="RecommendedProducts.noProducts" />
        </div>
      )}
    </div>
  );
};

RecommendedProductsComponent.defaultProps = {
  rootClassName: null,
  className: null,
  recommendedProducts: [],
  fetchRecommendedProductsInProgress: false,
  fetchRecommendedProductsError: null,
};

RecommendedProductsComponent.propTypes = {
  rootClassName: propTypes.string,
  className: propTypes.string,
  recommendedProductSKUs: propTypes.arrayOf(propTypes.string).isRequired,
  recommendedProducts: propTypes.arrayOf(propTypes.listing),
  fetchRecommendedProductsInProgress: propTypes.bool,
  fetchRecommendedProductsError: propTypes.error,
  onFetchRecommendedProducts: propTypes.func.isRequired,
};

const mapStateToProps = (state, ownProps) => {
  const {
    recommendedProductIds,
    fetchRecommendedProductsInProgress,
    fetchRecommendedProductsError,
  } = state.recommendedProducts || {};

  // Get products from marketplace data if we have the IDs
  const recommendedProducts = recommendedProductIds
    ? getMarketplaceEntities(
        state,
        recommendedProductIds.map(id => ({ id, type: 'listing' }))
      )
    : [];

  return {
    recommendedProducts,
    fetchRecommendedProductsInProgress,
    fetchRecommendedProductsError,
  };
};

const mapDispatchToProps = dispatch => ({
  onFetchRecommendedProducts: skus => dispatch(fetchRecommendedProducts(skus)),
});

const RecommendedProducts = compose(
  connect(mapStateToProps, mapDispatchToProps)
)(RecommendedProductsComponent);

export default RecommendedProducts;