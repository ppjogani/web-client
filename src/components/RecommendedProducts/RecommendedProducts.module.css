@import '../../styles/customMediaQueries.css';

.root {
  margin-top: 48px;
  margin-bottom: 24px;
  padding: 0 24px;

  @media (--viewportMedium) {
    margin-top: 72px;
    margin-bottom: 48px;
    padding: 0;
  }
}

.title {
  color: var(--colorGrey700);
  margin-bottom: 32px;
  text-align: center;

  @media (--viewportMedium) {
    margin-bottom: 48px;
  }
}

.productsGrid {
  display: grid;
  grid-template-columns: 1fr;
  gap: 24px;

  @media (--viewportMedium) {
    grid-template-columns: repeat(2, 1fr);
    gap: 32px;
  }

  @media (--viewportLarge) {
    grid-template-columns: repeat(3, 1fr);
    gap: 40px;
  }

  @media (--viewportXLarge) {
    grid-template-columns: repeat(4, 1fr);
  }
}

.productCard {
  text-decoration: none;
  color: inherit;
  border-radius: 8px;
  overflow: hidden;
  transition: transform 0.2s ease, box-shadow 0.2s ease;
  border: 1px solid var(--colorGrey100);

  &:hover {
    transform: translateY(-2px);
    box-shadow: 0 4px 12px rgba(0, 0, 0, 0.1);
    text-decoration: none;
    color: inherit;
  }
}

.imageContainer {
  position: relative;
  width: 100%;
  aspect-ratio: 4/3;
  background-color: var(--colorGrey50);
}

.productImage {
  width: 100%;
  height: 100%;
  object-fit: cover;
}

.noImage {
  display: flex;
  align-items: center;
  justify-content: center;
  width: 100%;
  height: 100%;
  background-color: var(--colorGrey100);
  color: var(--colorGrey500);
  font-size: 14px;
}

.productInfo {
  padding: 16px;

  @media (--viewportMedium) {
    padding: 20px;
  }
}

.productTitle {
  composes: h5 from global;
  margin: 0 0 8px 0;
  color: var(--colorGrey700);
  display: -webkit-box;
  -webkit-line-clamp: 2;
  -webkit-box-orient: vertical;
  overflow: hidden;
  line-height: 1.4;
}

.productPrice {
  composes: h4 from global;
  color: var(--colorGrey800);
  font-weight: var(--fontWeightSemiBold);
  margin: 0;
}

.loading,
.error,
.noProducts {
  text-align: center;
  padding: 48px 24px;
  color: var(--colorGrey500);

  @media (--viewportMedium) {
    padding: 72px 24px;
  }
}

.error {
  color: var(--colorFail);
}

.loading {
  color: var(--colorGrey600);
}