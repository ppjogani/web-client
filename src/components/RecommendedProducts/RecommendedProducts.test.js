import React from 'react';
import '@testing-library/jest-dom';
import { screen } from '@testing-library/react';
import { renderWithProviders } from '../../util/testHelpers';
import { createStore, combineReducers, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';

import { types as sdkTypes } from '../../util/sdkLoader';
import { createListing, createUser } from '../../util/testData';

import RecommendedProducts from './RecommendedProducts';
import recommendedProductsReducer from '../../ducks/recommendedProducts.duck';

const { Money } = sdkTypes;

const defaultMessages = {
  'RecommendedProducts.title': 'You may also like',
  'RecommendedProducts.heading': 'You may also like',
  'RecommendedProducts.loading': 'Loading recommended products...',
  'RecommendedProducts.loadError': 'Unable to load recommended products',
  'RecommendedProducts.noProducts': 'No recommended products available',
  'RecommendedProducts.noImage': 'No image available',
};

const createMockStore = (initialState = {}) => {
  const rootReducer = combineReducers({
    recommendedProducts: recommendedProductsReducer,
    marketplaceData: (state = { entities: {} }) => state,
  });

  return createStore(rootReducer, {
    recommendedProducts: {
      recommendedProductIds: null,
      fetchRecommendedProductsInProgress: false,
      fetchRecommendedProductsError: null,
      ...initialState,
    },
    marketplaceData: {
      entities: {},
    },
  }, applyMiddleware(thunk));
};

const createStoreWithProducts = (products = [], overrides = {}) => {
  const entities = {
    listing: {}
  };
  products.forEach(product => {
    entities.listing[product.id.uuid] = product;
  });

  const store = createMockStore({
    recommendedProductIds: products.map(p => p.id),
    ...overrides,
  });

  store.getState().marketplaceData.entities = entities;
  return store;
};

// Create a custom render function that uses the proper testHelpers renderWithProviders
// but still allows us to mock the dispatch for thunk actions
const renderWithStore = (component, store) => {
  const mockStore = store || createMockStore();

  // Override the dispatch method to capture thunk calls
  const originalDispatch = mockStore.dispatch;
  mockStore.dispatch = jest.fn(action => {
    if (typeof action === 'function') {
      // For thunk actions, call our mock and return a resolved promise
      mockFetchRecommendedProducts();
      return Promise.resolve();
    }
    return originalDispatch(action);
  });

  // Create initial state that matches what the proper store would have
  const state = mockStore.getState();

  return renderWithProviders(component, {
    initialState: state
  });
};

const createMockProduct = (sku, overrides = {}) => {
  return createListing(`listing-${sku}`, {
    title: `Product ${sku}`,
    description: `Description for product ${sku}`,
    price: new Money(2500, 'USD'),
    images: [
      {
        id: `image-${sku}`,
        type: 'image',
        attributes: {
          variants: {
            'listing-card': {
              url: `https://example.com/image-${sku}.jpg`,
              width: 400,
              height: 400,
            },
          },
        },
      },
    ],
    publicData: {
      sku: sku,
      brand: 'TestBrand',
      productUrl: `https://testbrand.com/product/${sku}`,
      listingType: 'sell-products',
      ...overrides.publicData,
    },
    ...overrides,
  }, {
    author: createUser(`user-${sku}`),
  });
};

// Mock the entire recommendedProducts duck to avoid SDK dependency
jest.mock('../../ducks/recommendedProducts.duck', () => ({
  __esModule: true,
  default: (state = { recommendedProductIds: null, fetchRecommendedProductsInProgress: false, fetchRecommendedProductsError: null }, action) => {
    switch (action.type) {
      case 'FETCH_RECOMMENDED_PRODUCTS_REQUEST':
        return { ...state, fetchRecommendedProductsInProgress: true, fetchRecommendedProductsError: null };
      case 'FETCH_RECOMMENDED_PRODUCTS_SUCCESS':
        return { ...state, fetchRecommendedProductsInProgress: false, recommendedProductIds: action.payload.productIds };
      case 'FETCH_RECOMMENDED_PRODUCTS_ERROR':
        return { ...state, fetchRecommendedProductsInProgress: false, fetchRecommendedProductsError: action.payload };
      default:
        return state;
    }
  },
  fetchRecommendedProducts: jest.fn(() => () => Promise.resolve()),
  FETCH_RECOMMENDED_PRODUCTS_REQUEST: 'FETCH_RECOMMENDED_PRODUCTS_REQUEST',
  FETCH_RECOMMENDED_PRODUCTS_SUCCESS: 'FETCH_RECOMMENDED_PRODUCTS_SUCCESS',
  FETCH_RECOMMENDED_PRODUCTS_ERROR: 'FETCH_RECOMMENDED_PRODUCTS_ERROR',
}));

// Get access to the mocked function
const mockFetchRecommendedProducts = require('../../ducks/recommendedProducts.duck').fetchRecommendedProducts;

describe('RecommendedProducts', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('renders nothing when no recommended SKUs provided', () => {
    const { container } = renderWithStore(
      <RecommendedProducts
        recommendedProductSKUs={[]}
      />
    );

    expect(container.firstChild).toBeNull();
  });

  it('renders nothing when recommendedSKUs is null or undefined', () => {
    const { container } = renderWithStore(
      <RecommendedProducts
        recommendedProductSKUs={null}
      />
    );

    expect(container.firstChild).toBeNull();
  });

  it('displays loading state', () => {
    const initialState = {
      recommendedProducts: {
        recommendedProductIds: null,
        fetchRecommendedProductsInProgress: true,
        fetchRecommendedProductsError: null,
      },
      marketplaceData: {
        entities: { listing: {} }
      }
    };

    renderWithProviders(
      <RecommendedProducts
        recommendedProductSKUs={['SKU001', 'SKU002']}
      />,
      { initialState }
    );

    expect(screen.getByText('Loading recommended products...')).toBeInTheDocument();
  });

  it('displays error state', () => {
    const initialState = {
      recommendedProducts: {
        recommendedProductIds: null,
        fetchRecommendedProductsInProgress: false,
        fetchRecommendedProductsError: { status: 500, statusText: 'Server Error' },
      },
      marketplaceData: {
        entities: { listing: {} }
      }
    };

    renderWithProviders(
      <RecommendedProducts
        recommendedProductSKUs={['SKU001', 'SKU002']}
      />,
      { initialState }
    );

    expect(screen.getByText('Unable to load recommended products')).toBeInTheDocument();
  });

  it('displays no products message when products array is empty', () => {
    const initialState = {
      recommendedProducts: {
        recommendedProductIds: [],
        fetchRecommendedProductsInProgress: false,
        fetchRecommendedProductsError: null,
      },
      marketplaceData: {
        entities: { listing: {} }
      }
    };

    renderWithProviders(
      <RecommendedProducts
        recommendedProductSKUs={['SKU001', 'SKU002']}
      />,
      { initialState }
    );

    expect(screen.getByText('No recommended products available')).toBeInTheDocument();
  });

  it('renders recommended products correctly', () => {
    const mockProducts = [
      createMockProduct('SKU001'),
      createMockProduct('SKU002'),
      createMockProduct('SKU003'),
    ];

    // Create initial state with products already loaded (no thunk call needed)
    const initialState = {
      recommendedProducts: {
        recommendedProductIds: mockProducts.map(p => p.id),
        fetchRecommendedProductsInProgress: false,
        fetchRecommendedProductsError: null,
      },
      marketplaceData: {
        entities: {
          listing: {}
        }
      }
    };

    // Add products to entities
    mockProducts.forEach(product => {
      initialState.marketplaceData.entities.listing[product.id.uuid] = product;
    });

    renderWithProviders(
      <RecommendedProducts
        recommendedProductSKUs={['SKU001', 'SKU002', 'SKU003']}
      />,
      { initialState }
    );

    expect(screen.getByText('You may also like')).toBeInTheDocument();
    expect(screen.getByText('Product SKU001')).toBeInTheDocument();
    expect(screen.getByText('Product SKU002')).toBeInTheDocument();
    expect(screen.getByText('Product SKU003')).toBeInTheDocument();
  });

  it('displays product prices correctly', () => {
    const mockProducts = [
      createMockProduct('SKU001', { price: new Money(1500, 'USD') }),
      createMockProduct('SKU002', { price: new Money(3000, 'USD') }),
    ];

    const store = createStoreWithProducts(mockProducts);

    renderWithStore(
      <RecommendedProducts
        recommendedProductSKUs={['SKU001', 'SKU002']}
      />,
      store
    );

    expect(screen.getByText('$15.00')).toBeInTheDocument();
    expect(screen.getByText('$30.00')).toBeInTheDocument();
  });

  it('renders product links to marketplace pages', () => {
    const mockProducts = [
      createMockProduct('SKU001'),
    ];

    const store = createMockStore({ products: mockProducts });

    renderWithStore(
      <RecommendedProducts
        recommendedProductSKUs={['SKU001']}
      />,
      store
    );

    const productLink = screen.getByRole('link');
    expect(productLink).toHaveAttribute('href', '/l/listing-sku001');
  });

  it('displays fallback when product image is missing', () => {
    const mockProducts = [
      createMockProduct('SKU001', { images: [] }),
    ];

    const store = createMockStore({ products: mockProducts });

    renderWithStore(
      <RecommendedProducts
        recommendedProductSKUs={['SKU001']}
      />,
      store
    );

    expect(screen.getByText('No image available')).toBeInTheDocument();
  });

  it('calls onFetchProducts with correct SKUs on mount', () => {
    const recommendedSKUs = ['SKU001', 'SKU002', 'SKU003'];

    renderWithStore(
      <RecommendedProducts
        recommendedProductSKUs={recommendedSKUs}
      />
    );

    expect(mockFetchRecommendedProducts).toHaveBeenCalled();
  });

  it('calls onFetchProducts when recommendedSKUs change', () => {
    const initialSKUs = ['SKU001', 'SKU002'];
    const newSKUs = ['SKU003', 'SKU004'];

    const { rerender } = renderWithStore(
      <RecommendedProducts
        recommendedProductSKUs={initialSKUs}
      />
    );

    expect(mockFetchRecommendedProducts).toHaveBeenCalled();

    rerender(
      <RecommendedProducts
        recommendedProductSKUs={newSKUs}
      />
    );

    expect(mockFetchRecommendedProducts).toHaveBeenCalledTimes(2);
  });

  it('does not call onFetchProducts when recommendedSKUs remain the same', () => {
    const recommendedSKUs = ['SKU001', 'SKU002'];

    const { rerender } = renderWithStore(
      <RecommendedProducts
        recommendedProductSKUs={recommendedSKUs}
      />
    );

    expect(mockFetchRecommendedProducts).toHaveBeenCalledTimes(1);

    // Re-render with same SKUs
    rerender(
      <RecommendedProducts
        recommendedProductSKUs={recommendedSKUs}
      />
    );

    // Should not call again
    expect(mockFetchRecommendedProducts).toHaveBeenCalledTimes(1);
  });

  it('handles products with missing brand information', () => {
    const mockProducts = [
      createMockProduct('SKU001', {
        publicData: { sku: 'SKU001', brand: undefined }
      }),
    ];

    const store = createMockStore({ products: mockProducts });

    renderWithStore(
      <RecommendedProducts
        recommendedProductSKUs={['SKU001']}
      />,
      store
    );

    expect(screen.getByText('Product SKU001')).toBeInTheDocument();
  });

  it('limits displayed products to maximum count', () => {
    const mockProducts = Array.from({ length: 10 }, (_, i) =>
      createMockProduct(`SKU${i.toString().padStart(3, '0')}`)
    );

    const store = createMockStore({ products: mockProducts });

    renderWithStore(
      <RecommendedProducts
        recommendedProductSKUs={mockProducts.map(p => p.attributes.publicData.sku)}
        maxProducts={4}
      />,
      store
    );

    // Should only display first 4 products
    expect(screen.getByText('Product SKU000')).toBeInTheDocument();
    expect(screen.getByText('Product SKU001')).toBeInTheDocument();
    expect(screen.getByText('Product SKU002')).toBeInTheDocument();
    expect(screen.getByText('Product SKU003')).toBeInTheDocument();
    expect(screen.queryByText('Product SKU004')).not.toBeInTheDocument();
  });

  it('matches snapshot with recommended products', () => {
    const mockProducts = [
      createMockProduct('SKU001'),
      createMockProduct('SKU002'),
    ];

    const store = createMockStore({ products: mockProducts });

    const { container } = renderWithStore(
      <RecommendedProducts
        recommendedProductSKUs={['SKU001', 'SKU002']}
      />,
      store
    );

    expect(container.firstChild).toMatchSnapshot();
  });
});