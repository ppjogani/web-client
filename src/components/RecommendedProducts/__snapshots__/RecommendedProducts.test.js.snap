// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecommendedProducts matches snapshot with recommended products 1`] = `null`;
